File: gpt_request.py
import os
from dotenv import load_dotenv

import requests
import threading

load_dotenv()

API_KEY = os.getenv("OPENROUTER_API_KEY")
MODEL = "mistralai/mistral-7b-instruct:free"
MODEL1 = "google/gemma-2-9b-it:free"
list = [None, None] # To store results from both models

def expand_sentence(prompt_text, model, index):
    url = "https://openrouter.ai/api/v1/chat/completions"

    headers = {
        "Authorization": f"Bearer {API_KEY}",
        "Content-Type": "application/json",
        "HTTP-Referer": "http://localhost",     
        "X-Title": "EliteBallTranslator"
    }

    payload = {
        "model": model,
        "messages": [
            {"role": "system", "content": "You are a writing assistant."},
            {"role": "user", "content": f"Summarize this sentence professionally using as few words as possible: {prompt_text}"}
        ],
        "temperature": 0.4,
        "max_tokens": 100
    }

    response = requests.post(url, headers=headers, json=payload)

    print("Status Code:", response.status_code)
    print("Response Text:", response.text)

    response.raise_for_status()
    data = response.json()
    data = data["choices"][0]["message"]["content"].strip()

    cleaned_data = clean_output(data)
    
    list[index] = cleaned_data
    return cleaned_data

def clean_output(text):
    
    # Entferne häufige Prompt-Tokens
    unwanted_patterns = [
        r'<s>', r'</s>', r'\[INST\]', r'\[/INST\]', 
        r'\[OUT\]', r'\[/OUT\]', r'<\|.*?\|>',
        r'^[^a-zA-Z0-9]*', r'[^a-zA-Z0-9\.!?]*$' 
    ]
    
    cleaned_text = text
    for pattern in unwanted_patterns:
        cleaned_text = re.sub(pattern, '', cleaned_text)
    
    # Entferne überflüssige Leerzeichen und trimme
    cleaned_text = ' '.join(cleaned_text.split()).strip()
    
    # Stelle sicher, dass der Text mit einem Buchstaben oder Zahl beginnt
    if cleaned_text and not cleaned_text[0].isalnum():
        cleaned_text = cleaned_text[1:].strip()
    
    return cleaned_text if cleaned_text else text 

def expand_sentence_butThreaded(prompt_text):

    thread1 = threading.Thread(target=expand_sentence, args=(prompt_text, MODEL, 0))
    thread2 = threading.Thread(target=expand_sentence, args=(prompt_text, MODEL1, 1))
    thread1.start()
    thread2.start()

    thread1.join()
    thread2.join()
    return checkList()

def checkList():
    if list[0] is None:
        return list[1]
    else:
        return list[0]

File: frontend/lib/gptRequest.ts
const API_KEY = process.env.OPENROUTER_API_KEY;
const MODEL1 = "google/gemma-3n-e2b-it:free";
const MODEL2 = "google/gemma-3n-e4b-it:free";

async function expandSentence(promptText: string, model: string): Promise<string> {
  const url = "https://openrouter.ai/api/v1/chat/completions";

  const headers = {
    "Authorization": `Bearer ${API_KEY}`,
    "Content-Type": "application/json",
    "HTTP-Referer": "https://eliteballknowledge.vercel.app",
    "X-Title": "EliteBallTranslator"
  };

  const payload = {
    "model": model,
    "messages": [
      {"role": "user", "content": `Rewrite this phrase into one concise and professional sentence, without giving me extra infromation: "${promptText}"`}
    ],
    "temperature": 0.7,  
    "max_tokens": 80    
  };

  console.log(`Calling OpenRouter API with model: ${model}`);

  const response = await fetch(url, {
    method: 'POST',
    headers: headers,
    body: JSON.stringify(payload)
  });

  if (!response.ok) {
    throw new Error(`HTTP error! status: ${response.status}`);
  }

  const data = await response.json();

  let text = data.choices[0].message.content.trim();
  return cleanOutput(text);
}

function cleanOutput(text: string): string {
  const unwantedPatterns = [
    /<s>/g, /<\/s>/g, /\[INST\]/g, /\[\/INST\]/g, 
    /\[OUT\]/g, /\[\/OUT\]/g, /<\|.*?\|>/g,
    /^[^a-zA-Z0-9]*/, /[^a-zA-Z0-9\.!?]*$/ 
  ];
  
  let cleanedText = text;
  unwantedPatterns.forEach(pattern => {
    cleanedText = cleanedText.replace(pattern, '');
  });
  
  cleanedText = cleanedText.split(/\s+/).join(' ').trim();
  
  if (cleanedText && !cleanedText[0].match(/[a-zA-Z0-9]/)) {
    cleanedText = cleanedText.slice(1).trim();
  }
  
  return cleanedText || text;
}

export async function expandSentenceButThreaded(promptText: string): Promise<string> {
  console.log("API Key vorhanden:", !!API_KEY);
  console.log("API Key Länge:", API_KEY?.length);

  try {
    // Versuche zuerst MODEL1, falls das fehlschlägt, versuche MODEL2
    try {
      const result1 = await expandSentence(promptText, MODEL1);
      return result1;
    } catch (error1) {
      console.log(`Model ${MODEL1} failed, trying ${MODEL2}:`, error1);
      const result2 = await expandSentence(promptText, MODEL2);
      return result2;
    }
  } catch (error) {
    console.error("Both models failed in expandSentenceButThreaded:", error);
    throw error;
  }
}